projects/RabbitMQ.Client/client/impl/Connection.cs

```diff
@@ -871,7 +877,12 @@ public void HeartbeatWriteTimerCallback(object state)
             {
                 if (!_closed)
                 {
-                    Write(Client.Impl.Framing.Heartbeat.GetHeartbeatFrame());
+                    _writeLock.Wait();
+                    Span<byte> span = _frameHandler.PipeWriter.GetSpan(Client.Impl.Framing.Heartbeat.Payload.Length);
+                    Client.Impl.Framing.Heartbeat.Payload.CopyTo(span);
+                    _frameHandler.PipeWriter.Advance(Client.Impl.Framing.Heartbeat.Payload.Length);
+                    _frameHandler.Flush();
```

Comment:
you should add try finally to the locking in case any method throws. this ofcourse applies to all places where the lock is used.

Reply:
yeah, I've noticed some room for improvements after the profiling as well, will make some changes today :) Going to try a few scenarios.